cmd/skopeo/proxy.go

```diff
@@ -232,6 +233,56 @@ func (h *proxyHandler) OpenImage(args []interface{}) (replyBuf, error) {
 	return ret, nil
 }
 
+func isNotFoundImageError(err error) bool {
+	msg := err.Error()
+	// TODO: Drive a correct API for this into containers/image
+	return strings.HasSuffix(msg, "manifest unknown") ||
+		strings.Contains(msg, ": no descriptor found for reference")
```

Comment:
Not a full review, just a quick note for now:

You’re right, there isn‘t a good abstraction for this (primarily because the registries behave quite differently and we just didn’t do the work to make a survey).

Currently, I’d recommend searching for `docker/distribution` error types, similarly to https://github.com/containers/image/blob/62be088831a0836823930f0c226152b164ff4241/docker/docker_client.go#L984 . Precedent for making that assumption outside of c/image: https://github.com/containers/common/blob/fd9763624b9c81d128589058e05460323988bf53/pkg/retry/retry.go#L52 .

But, also, note the changes to `isManifestUnknownError` proposed in https://github.com/containers/image/pull/1299 . So the integration test is quite necessary to keep such code (whether it relies on string matching or undocumented d/distribution type usage) working.

Reply:
> Currently, I’d recommend searching for docker/distribution error types, 

That'd pull a direct dependency on docker/distribution into skopeo; but I dunno, perhaps that doesn't matter.

Maybe a local minima of effort/hackiness is `s/isManifestUnknownError/BestEffortIsManifestUnknownError/` in c/image (i.e. make it public but with documented caveats)?

In the c/image OCI case, we clearly can add a reliable API for this.  I can perhaps look at that, just trying to solve a higher level problem at the moment.

Reply:
A direct dependency (on the _error definitions_, not all of d/distribution) is fine; we already have the indirect one.

---

WRT making `isManifestUnknownError` public:

(Note that function is more or less _known_ not to be sufficient. In the worst case, IIRC some registries (in some, possibly very old, versions?) return a HTTP 404 without the API-expected response at all. c/image/docker only uses it behind an opt-in, so that users with unhandled registries can still get the other functionality. That generally applies to the idea of this PR — callers can’t assume it will _reliably_ report “no error, image not present”.[1])

Defining the c/image API, and for not documenting it as a best-effort thing, is a good idea — not something I’d block this PR on, but more valuable than either of the hacks discussed so far.

Just one thing: I’d prefer the API to be closer to the final shape, `errors.As(err, c/image/docker.SomeNewErrorType)` rather than a function.


[1] Note that “no error, image not present” is somewhat ill-defined. At least in the past, the Docker Hub would report “permission denied” on references to any missing repo, on the theory that users who don’t have the permission to view a repo don’t have the permission to know whether it exists.

---

If you are talking about the OCI format on registries, I don’t think that pulling v2s2 vs. pulling OCI makes any difference. If you are talking about the `oci` and `oci-archive` formats, yes, we clearly can add a reliable API in that case.